common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+    {
+    }
+};
+
+class UdpServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit UdpServerSocket(const std::shared_ptr<UdpSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdpClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit UdpClientSocket(const std::shared_ptr<UdpSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class TcpServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit TcpServerSocket(const std::shared_ptr<TcpSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class TcpClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit TcpClientSocket(const std::shared_ptr<TcpSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdsServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit UdsServerSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class UdsClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit UdsClientSocket(const std::shared_ptr<UdsSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+/**
+ * prplmesh-democracy: Do you prefer templates?
+ * These template classes would replace previous ServerSocket and ClientSocket derived classes
+ */
+template <class SocketType> class ServerSocketImpl : public ServerSocketAbstractImpl {
+public:
+    explicit ServerSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+template <class SocketType> class ClientSocketImpl : public ClientSocketAbstractImpl {
+public:
+    explicit ClientSocketImpl(const std::shared_ptr<SocketType> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};
+
+} // namespace sockets
+
+/**
+ * A CMDU handler class is the equivalent in functionality to, for example, each of those methods
+ * found in the backhaul manager that are used to handle the different types of messages.
+ *
+ * To test a particular implementation of CMDU handler, we must know which are its dependencies and
+ * what does the handler do with them. All dependencies must be injected as parameters of the
+ * handler constructor and the handler code must be programmed to interfaces, not to particular
+ * implementations. A dependency that most if not all CMDU handlers will have is the socket
+ * connection that the message was received on and that will be used to send the CMDU replies.
+ * This way we can write tests for each implementation of the CMDU handler just by using mocks of
+ * the dependencies and setting expectations on the mocks, that is, without sending any single
+ * byte on a real socket connection.
+ */
+class CmduHandler {
+public:
+    virtual ~CmduHandler()                                          = default;
+    virtual void handle_message(ieee1905_1::CmduMessageRx &cmdu_rx) = 0;

```

Comment:
I think it would be useful to make this templated. We will want to use this (and the other classes) as well for WSC attributes, beerocks messages, etc.